source/utilities.tex

```diff
@@ -11504,6 +11545,462 @@
+\tcode{s.reset(static_cast<SP>(p), std::forward<Args>(args)...)}
+is well-\newline formed;
+\item
+otherwise,
+\begin{codeblock}
+if (p) {
+  apply([&](auto&&... args) {
+    s = Smart(static_cast<SP>(p), std::forward<Args>(args)... ); }, std::move(a));
+}
+\end{codeblock}
+if \tcode{is_constructible_v<Smart, SP, Args...>} is \tcode{true};
+\item
+otherwise, the program is ill-formed.
+\end{itemize}
+\end{itemdescr}
+
+\begin{itemdecl}
+operator Pointer*() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\expects
+\tcode{operator void**()} has not been called on \tcode{*this}.
+
+\pnum
+\returns
+\tcode{addressof(const_cast<Pointer\&>(p))}.
+\end{itemdescr}
+
+\begin{itemdecl}
+operator void**() const noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{is_same_v<Pointer, void*>} is \tcode{false}.
+
+\pnum
+\mandates
+\tcode{is_pointer_v<Pointer>} is \tcode{true}.
+
+\pnum
+\expects
+\tcode{operator Pointer*()} has not been called on \tcode{*this}.
+
+\pnum
+\returns
+A pointer value \tcode{v} such that:
+\begin{itemize}
+\item
+the initial value \tcode{*v} is equivalent to \tcode{static_cast<void*>(p)} and
+\item
+any modification of \tcode{*v}
+that is not followed by a subsequent modification of \tcode{*this}
+affects the value of \tcode{p} during the destruction of \tcode{*this},
+such that \tcode{static_cast<void*>(p) == *v}.
+\end{itemize}
+
+\pnum
+\remarks
+Accessing \tcode{*v} outside the lifetime of \tcode{*this}
+has undefined behavior.
+
+\pnum
+\begin{note}
+\tcode{reinterpret_cast<void**>(static_cast<Pointer*>(*this))}
+can be a viable implementation strategy.
+\end{note}
+\end{itemdescr}
+
+\rSec2[out.ptr]{Function template \tcode{out_ptr}}
+
+\indexlibraryglobal{out_ptr}%
+\begin{itemdecl}
+template<class Pointer = void, class Smart, class... Args>
+  auto out_ptr(Smart& s, Args&&... args);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+Let \tcode{P} be \tcode{Pointer}
+if \tcode{is_void_v<Pointer>} is \tcode{false},
+otherwise \tcode{\exposid{POINTER_OF}(Smart)}.
+
+\pnum
+\returns
+\tcode{out_ptr_t<Smart, P, Args\&\&...>(s, std::forward<Args>(args)...)}
+\end{itemdescr}
+
+\rSec2[inout.ptr.t]{Class template \tcode{inout_ptr_t}}
+
+\pnum
+\tcode{inout_ptr_t} is a class template used types
+such as smart pointers\iref{smartptr},
+with functions that use output pointer parameters
+whose dereferenced values may first be deleted
+before being set to another allocated value.
+
+\pnum
+\begin{example}
+\begin{codeblock}
+#include <memory>
+
+struct StarFish* starfish_alloc();
+int starfish_populate(struct StarFish** ps, const char *description);
+
+struct StarFishDeleter {
+  void operator() (struct StarFish* c) const noexcept;
+};
+
+using StarFishPtr = std::unique_ptr<StarFish, StarFishDeleter>;
+
+int main(int, char*[]) {
+  StarFishPtr peach(starfish_context());
+  // ...
+  // used, need to re-make
+  int err = starfish_populate(std::inout_ptr(peach), "caring clown-fish liker");
+  return err;
```

Comment:
P1132R8 changes this example to use `snake_case` not `PascalCase` (and the `*` is misplaced in `star_fish_populate` even in R8):

```suggestion
struct star_fish* star_fish_alloc();
int star_fish_populate(struct star_fish** ps, const char* description);

struct star_fish_deleter {
  void operator() (struct star_fish* c) const noexcept;
};

using star_fish_ptr = std::unique_ptr<star_fish, star_fish_deleter>;

int main(int, char*[]) {
  star_fish_ptr peach(star_fish_context());
  // ...
  // used, need to re-make
  int err = star_fish_populate(std::inout_ptr(peach), "caring clown-fish liker");
  return err;
```